File: ChainOfResponsibility/ChainOfResponsibilityGUI.py
from abc import ABC, abstractmethod

class Handler(ABC):
    def __init__(self, parent=None):
        self._parent = parent
        self._helpText = None

    @abstractmethod
    def show_helper_text(self):
        pass

    def set_helper_text(self, text):
        self._helpText = text



class Container(Handler):
    def show_helper_text(self):
        if self._helpText is not None:  # Can handle request.
            print("Help :: ", self._helpText)
        elif self._parent is not None:
            print("Message passed to next in chain by Container")
            self._parent.show_helper_text()



class Button(Handler):
    def __init__(self, label, parent=None ):
        super(Button, self).__init__(parent)
        self._label = label

    def show_helper_text(self):
        if self._helpText is not None:  # Can handle request.
            print("Help :: ", self._helpText)
        elif self._parent is not None:
            print("Message passed to next in chain by Button")
            self._parent.show_helper_text()


class Panel(Handler):
    def show_helper_text(self):
        if self._helpText is not None:  # Can handle request.
            print("Help :: ", self._helpText)
        elif self._parent is not None:
            print("Message passed to next in chain by Panel")
            self._parent.show_helper_text()


# Client Code
p = Panel()
p.set_helper_text("Panel help text.")
b1 = Button("Ok", p)
b1.set_helper_text("Ok button help text.")
b2 = Button("Cancel", p)
b1.show_helper_text()
b2.show_helper_text()

"""
Help ::  Ok button help text.
Message passed to next in chain by Button
Help ::  Panel help text.
"""

File: ChainOfResponsibility/ChainOfResponsibilityGUI.java
abstract class Handler {
    protected Handler parent;
    protected String helpText;

    public Handler(Handler parent) {
        this.parent = parent;
        this.helpText = null;
    }

    public abstract void showHelperText();

    public void setHelperText(String text) {
        this.helpText = text;
    }
}

class Container extends Handler {
    public Container(Handler parent) {
        super(parent);
    }

    @Override
    public void showHelperText() {
        if (helpText != null) {
            System.out.println("Help :: " + helpText);
        } else if (parent != null) {
            System.out.println("Message passed to next in chain by Container");
            parent.showHelperText();
        }
    }
}

class Button extends Handler {
    private String label;

    public Button(String label, Handler parent) {
        super(parent);
        this.label = label;
    }

    @Override
    public void showHelperText() {
        if (helpText != null) {
            System.out.println("Help :: " + helpText);
        } else if (parent != null) {
            System.out.println("Message passed to next in chain by Button");
            parent.showHelperText();
        }
    }
}

class Panel extends Handler {
    public Panel() {
        super(null);
    }

    @Override
    public void showHelperText() {
        if (helpText != null) {
            System.out.println("Help :: " + helpText);
        } else if (parent != null) {
            System.out.println("Message passed to next in chain by Panel");
            parent.showHelperText();
        }
    }
}

public class ChainOfResponsibilityGUI {
    public static void main(String[] args) {
        Panel p = new Panel();
        p.setHelperText("Panel help text.");

        Button b1 = new Button("Ok", p);
        b1.setHelperText("Ok button help text.");

        Button b2 = new Button("Cancel", p);

        b1.showHelperText();
        b2.showHelperText();
    }
}

/*
Help :: Ok button help text.
Message passed to next in chain by Button
Help :: Panel help text.
*/